### DISCUSSION TITLE
What does wasm-bindgen-test test?

### DISCUSSION THREAD

**Discussion by daidoji** [wasm-bindgen/wasm-bindgen#3744] [Q&A] [answered]
Sorry, I'm a little confused.  I've gotten through the game of life tutorial and read the wasm-bindgen book and even integrated a project into a usable wasm js-callable library.  However, my mental models of what wasm-bindgen are doing and where to do what came up against this question today as I was trying to write tests.

Question 1.
If wasm-bindgen is taking (Rust) -> (wasm) which can then be called from a js library.  Why then when I write a wasm-bindgen-test I'm testing the Rust interface that wasm-bindgen is using to generate the wasm?  Wouldn't it be better to test the wasm artifact itself? 

Question 2.
Then in a matter of logistics if I have a Crate C, a wasm-wrapper for that crate W, and a javascript application A I'm trying to use W within, how do people manage testing and where/what do you test?

In my naive approach I thought:
1. rstest for the Crate C
2. wasm-bindgen-test for the wasm-artifact W
3. application tests/startup checks for A to make sure the library is integrating properly in javascript somewhere.

However, based on my first question it looks like I'm just writing more Rust code against the Rust wrapper in W for 2 and the wasm-artifact itself is never getting tested.  What am I missing and how do other professionals do it?  Any clarification would be helpful.

**Comment by daxpedda** [comment, marked as answer]
> Question 1.
> If wasm-bindgen is taking (Rust) -> (wasm) which can then be called from a js library. Why then when I write a wasm-bindgen-test I'm testing the Rust interface that wasm-bindgen is using to generate the wasm? Wouldn't it be better to test the wasm artifact itself?

`wasm-bindgen-cli` is just a Wasm post-processor. The Rust compiler is the one converting Rust to Wasm. `wasm-bindgen-test` + `wasm-bindgen-test-runner` executes your tests when targeting `wasm32-unknown-unknown` on a platform that supports Wasm (e.g. a browser), so it does indeed test how your code runs as Wasm.

> Question 2.
> Then in a matter of logistics if I have a Crate C, a wasm-wrapper for that crate W, and a javascript application A I'm trying to use W within, how do people manage testing and where/what do you test?

I don't completely understand your question, also never heard of `rstest`. Also not sure what a "wasm-wrapper" is if "Crate C" is already written in Rust.

Generally speaking you can use the regular Cargo test infrastructure in combination with `wasm-bindgen-test` + `wasm-bindgen-test-runner` to test your code **as Wasm**.
If you want to test your JS API (that you expose through `wasm-bindgen`), you should test it with whatever the JS ecosystem offers (I'm not familiar with it).

**Reply by daidoji** [reply]
Ahh thank you.  Sorry rstest is a fixture tester for Rust.  I think I understand now.  

My second question was maybe more of a subjective one.  Like when you're making a rust crate C with #[wasm_bindgen_macros] say that'll expose some interface to my js application.  I was wondering what types of things people tested where?  

For example: 
- unit tests on the crate
- integration tests or interface of the wasm artifact in wasm-bindgen-test tests
- runtime/startup/application integration tests in the js application

Or are there other things that people do to manage the fact that they might like to distribute the wasm-artifact to a bunch of different places like a library.

**Reply by daxpedda** [reply]
This is probably similar to whatever you would do in any other Rust crate.
- Unit tests should live next to the code it's testing.
- Integration tests should live at the crate its testing.

Personally I only use Rust in the browser without any external JS application, so I have no experience in how to do integration tests from the JS side.

What I personally would do in your case, is use unit tests wherever possible/needed and test "crate C" with integration tests separately from your "crate W".

I maintain a very small crate [`web-time`](https://github.com/daxpedda/web-time), where I have both unit and integration tests. I hope this can help as some form of inspiration.
